### PR TITLE
Shifting message parsing responsibility to container runtime

### PR DESCRIPTION
**Pull request by vladsud**
This change prepares code base to shift responsibility for message content parsing from loader to runtime layer.
This helps for two kinds of processes:
- compressed payloads - they are base64 strings now, binary payload latter. No need to wrap with JSON
- chunked ops - wrapping with JSON quite often doubles payload size due to escaping.

It also lets all ops go to runtime, preparing code base for removal of cross-layer reaching. I.e. in future releases we will remove ability of runtime to reach out to DeltaManager "op" handler.
Runtime is in the right position to expose proper information as it sees fit (specifically this applies to summary-related ops).

Validated that runtime can parse ops just fine if Loader stops parsing. This change will be made in next branch for back-compat reasons.

Please note that "op" handler in runtime now has extra arg to indicate if it's a runtime message or not.
For now, for back-compat reasons, it get invoked only for runtime messages, but that's going to change in the future.

Opened the following bugs to track issues discovered in Pending State manager and dirty flag management:
- https://dev.azure.com/fluidframework/internal/_workitems/edit/1832
- https://dev.azure.com/fluidframework/internal/_workitems/edit/1833

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -1546,6 +1553,9 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
             this._perfSignalData.signalsLost = 0;
             this._perfSignalData.signalTimestamp = 0;
             this._perfSignalData.trackingSignalSequenceNumber = undefined;
+        } else {
+            assert(this.attachState === AttachState.Attached,
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1557]: should this change be part of this PR?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -529,13 +535,14 @@ export function unpackRuntimeMessage(message: ISequencedDocumentMessage) {
             message.type = innerContents.type;
             message.contents = innerContents.contents;
         }
-        assert(isUnpackedRuntimeMessage(message), 0x122 /* "Message to unpack is not proper runtime message" */);
+        return true;
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:538]: #nit this function is kind of weird because it mutates the parameter as a side effect and then returns true/false if the message is an op. Maybe the boolean return can be a separate function or just a comparison of the `message.type`.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -2473,6 +2492,14 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
     }
 
     private updateDocumentDirtyState(dirty: boolean) {
+        if (this.attachState !== AttachState.Attached) {
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:2495]: This seems unrelated

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 4]
```diff
@@ -233,7 +233,11 @@ export class PendingStateManager implements IDisposable {
 
             // then we push onto pendingStates which will cause PendingStateManager to resubmit when we connect
             // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-            this.pendingStates.push(this.initialStates.shift()!);
+            const message = this.initialStates.shift()!;
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:236]: same for this

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1557]: Some tests failed with this change, and they were incorrectly "hacking" container runtime, so it was easier to catch wrongdoing by adding more checks like that. I can move it out, but 1-liners do not feel worth the effort :)

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:538]: The verb in name "unpack" tells what it does, so I think it's fine from that POV. We already have that other function - isRuntimeMessage(), if you feel strong about it, I can change it, but all the places that call this API would need to call both functions, that seems like a waste to me

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:2495]: Yes. You can help me get rid of it from this PR :) - it exists as stand-alone fixes - https://github.com/microsoft/FluidFramework/pull/11823.
It's required for this change to succeed (indirectly, as some tests were failing due to bugs, and these asserts help discover these bugs and prevent them in the future).

File: packages/runtime/container-runtime/src/pendingStateManager.ts (hunk 4 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:236]: Same as above - these are bugs that existed in the code and my changes uncovered them - exists as stand-alone PR https://github.com/microsoft/FluidFramework/pull/11823 that should be merged before this change

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1557]: Actually, it's also part of https://github.com/microsoft/FluidFramework/pull/11823 :)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 5]
```diff
@@ -1589,40 +1596,40 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
+        if (typeof message.contents === "string" && message.contents !== "") {
+            message.contents = JSON.parse(message.contents);
+        }
+
+        // Caveat: This will return false for runtime message in very old format, that are used in snapshot tests
+        // This format was not shipped to production workflows.
+        const runtimeMessage = unpackRuntimeMessage(message);
+
+        if (this.mc.config.getBoolean("enableOfflineLoad") ?? this.runtimeOptions.enableOfflineLoad) {
+            this.savedOps.push(messageArg);
```
Comment (NicholasCouri) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1617]: >messageArg [](http://example.com/codeflow?start=31&length=10)

do we need to push messageArg or message to the saveOps list ?

Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1617]: messageArg is the right thing, as we push all ops (not just runtime ops). Take a look at processSavedOps() - it feeds these same ops back to process(), so whatever we get here is what we want to save

This change prepares code for receiving unprocessed ops, but no change is done in loader layer to stop processing ops.
That change will be done in next. I need at least one release to ensure that new loader can work with "old" runtime (where "old" is this PR).

File: api-report/container-runtime.api.md [hunk 6]
```diff
@@ -768,7 +768,7 @@ export class SummaryCollection extends TypedEventEmitter<ISummaryCollectionOpEve
 }
 
 // @public (undocumented)
-export function unpackRuntimeMessage(message: ISequencedDocumentMessage): ISequencedDocumentMessage;
+export function unpackRuntimeMessage(message: ISequencedDocumentMessage): boolean;
```
[the remark this answers is not in the document]
Comment (vladsud) [reply api-report/container-runtime.api.md:771]: These are all internal APIs. They are exported from packages as we need them in other packages, but we need to come up with some kind of "internal" mark-up. I believe do not have a design here.
Glancing through the work items, I'm not sure we have consensus here:
https://dev.azure.com/fluidframework/internal/_workitems/edit/1774
https://dev.azure.com/fluidframework/internal/_workitems/edit/1775
https://dev.azure.com/fluidframework/internal/_workitems/edit/1776

File: packages/runtime/runtime-definitions/src/dataStoreContext.ts [hunk 7]
```diff
@@ -84,8 +84,9 @@ export const VisibilityState = {
 };
 export type VisibilityState = typeof VisibilityState[keyof typeof VisibilityState];
 
-export interface IContainerRuntimeBaseEvents extends IEvent {
-    (event: "batchBegin" | "op", listener: (op: ISequencedDocumentMessage) => void);
+export interface IContainerRuntimeBaseEvents extends IEvent{
+    (event: "batchBegin", listener: (op: ISequencedDocumentMessage) => void);
+    (event: "op", listener: (op: ISequencedDocumentMessage, runtimeMessage?: boolean) => void);
```
[the remark this answers is not in the document]
Comment (vladsud) [reply packages/runtime/runtime-definitions/src/dataStoreContext.ts:89]: Correct, it's not a breaking change, though I might want to mention it such that people start changing their code

File: packages/test/test-end-to-end-tests/src/test/batching.spec.ts [hunk 8]
```diff
@@ -40,8 +40,8 @@ async function yieldJSTurn(): Promise<void> {
 }
 
 function setupBatchMessageListener(dataStore: ITestFluidObject, receivedMessages: ISequencedDocumentMessage[]) {
-    dataStore.context.containerRuntime.on("op", (message: ISequencedDocumentMessage) => {
-        if (isRuntimeMessage(message)) {
+    dataStore.context.containerRuntime.on("op", (message: ISequencedDocumentMessage, runtimeMessage?: boolean) => {
+        if (runtimeMessage !== false) {
```
[the remark this answers is not in the document]
Comment (vladsud) [reply packages/test/test-end-to-end-tests/src/test/batching.spec.ts:44]: Right, it's done in a way where we do not need to make changes to this line in the future (or at least do not have too). I.e. if you deal with old container runtime, then you are getting undefined here, and if you are using new runtime, then you are getting true/false. The API shape is done in a way not to be breaking, but I agree it adds confusion - I'll add some comments around.

[the remark this answers is not in the document]
Comment (vladsud) [reply packages/test/test-end-to-end-tests/src/test/batching.spec.ts:44]: So I've added comment in event signature. Will add breaking.md notes (it's not really breaking - more FYI) in separate PR as I have to refactor breaking.md sections - we screwed it up, so I do not want to do so in this PR.

File: api-report/container-runtime.api.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (vladsud) [reply api-report/container-runtime.api.md:771]: Added @internal and some description

**Comment by vladsud** [comment]
Ping - would appreciate if I can get it in.

**Review by jatgarg — APPROVED** [review]

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 5 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1617]: > Take a look at processSavedOps() - it feeds these same ops back to process(),

Yes, I had seen and misunderstood it - now it makes sense.

**Review by NicholasCouri — APPROVED** [review]
:shipit:

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (alexvy86) [reply packages/runtime/container-runtime/src/containerRuntime.ts:545]: I think at least adding function docs that make it clear that the parameter is mutated in-place (bonus points if it describes *how* it changes), is worth it.

Comment (alexvy86) [reply packages/runtime/container-runtime/src/containerRuntime.ts:545]: And now that I think about it, also what the return value means.

The alternative of returning the unpacked message and keeping the parameter untouched does feel more friendly to new programmers (less bug prone) but we might not want to have to allocate to clone the message. I think FF in general is complex enough that I don't mind going with the current approach.

**Review by andre4i — APPROVED** [review]
